src/hidimstat/gaussian_knockoff.py

```diff
@@ -1,169 +1,179 @@
+    using the equi-correlated method described in :cite:`barber2015controlling` and
+    :cite:`candes2018panning`. The function takes as input the design matrix X,
+    the vector of empirical mean values mu, and the empirical covariance
+    matrix sigma. It returns the knockoff variables X_tilde.
+
+    The original implementation can be found at
+    https://github.com/msesia/knockoff-filter/blob/master/R/knockoff/R/create_gaussian.R
 
     Parameters
     ----------
     X: 2D ndarray (n_samples, n_features)
-        original design matrix
+        The original design matrix.
 
     mu : 1D ndarray (n_features, )
-        vector of empirical mean values
+        A vector of empirical mean values.
+
+    sigma : 2D ndarray (n_samples, n_features)
+        The empirical covariance matrix.
+
+    seed : int, optional
+        A random seed for generating the uniform noise used to create the knockoff variables.
 
-    method: str
-        method to generate gaussian knockoff
+    tol : float, default=1.e-14
+        A tolerance value used for numerical stability in the calculation of the Cholesky decomposition.
 
-    Sigma : 2D ndarray (n_samples, n_features)
-        empirical covariance matrix
+    repeat : bool, default=False
+        If True, the function returns the values used to generate the knockoff variables (mu_tilde and sigma_tilde_decompose),
+        which can be used to generate additional knockoff variables without having to recompute them.
 
     Returns
     -------
     X_tilde : 2D ndarray (n_samples, n_features)
-        knockoff design matrix
-    """
-    memory = check_memory(memory)
+        The knockoff variables.
 
+    References
+    ----------
+    .. footbibliography::
+    """
     n_samples, n_features = X.shape
-    if method == "equi":
-        Diag_s = np.diag(_s_equi(Sigma))
-    else:
-        raise ValueError(
-            "{} is not a valid knockoff " "contriction method".format(method)
-        )
 
-    Sigma_inv_s = np.linalg.solve(Sigma, Diag_s)
+    # create a uniform noise for all the data
+    rng = np.random.RandomState(seed)
+    u_tilde = rng.randn(n_samples, n_features)
+
+    diag_s = np.diag(_s_equi(sigma, tol=tol))
 
-    # First part on the RHS of equation 1.4 in Barber & Candes (2015)
-    Mu_tilde = X - np.dot(X - mu, Sigma_inv_s)
+    sigma_inv_s = np.linalg.solve(sigma, diag_s)
+
+    # First part on the RHS of equation 1.4 in barber2015controlling
+    mu_tilde = X - np.dot(X - mu, sigma_inv_s)
     # To calculate the Cholesky decomposition later on
-    Sigma_tilde = 2 * Diag_s - Diag_s.dot(Sigma_inv_s.dot(Diag_s))
-    while not _is_posdef(Sigma_tilde):
-        Sigma_tilde += 1e-10 * np.eye(n_features)
+    sigma_tilde = 2 * diag_s - diag_s.dot(sigma_inv_s)
+    # test is the matrix is positive definite
+    while not np.all(np.linalg.eigvalsh(sigma_tilde) > tol):
+        sigma_tilde += 1e-10 * np.eye(n_features)
```

Comment:
This might be expensive. I think that having the increment fixed and pre-defined may not be optimal. 
In PyRiemann, they have a more principled way for finding the nearest SPD matrix, using bisections, see: https://github.com/pyRiemann/pyRiemann/blob/409869629642f6f860417992090a8127c4a0d255/pyriemann/utils/base.py#L160-L200

Reply:
I propose to keep that fior later, hence I opened issue #150